File: 668-kth-smallest-number-in-multiplication-table/kth-smallest-number-in-multiplication-table.cpp
class Solution {
public:
bool check(int x,int m,int n,int k)

{
    int ans=0;
    for(int i=1;i<=m;i++)
    {
        int a=x/i;
        ans+=min(a,n);
    }
    if(ans>=k)
    return true;
    else
    return false;
}
    int findKthNumber(int m, int n, int k) {
        int low=1;
        int high=m*n;
        int ans=-1;
        while(low<=high)
        {
            int mid=low+(high-low)/2;
            if(check(mid,m,n,k))
            {
                ans=mid;
                high=mid-1;
            }
            else
            low=mid+1;
        }
        return ans;
    }
};

File: 668-kth-smallest-number-in-multiplication-table/kth-smallest-number-in-multiplication-table.py
class Solution:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        def count(x:int):
            cnt=0
            for i in range(m):
                mul=i+1
                beg=mul
                end=(i+1)*n
                max_idx=x//mul
                cnt+=min(max_idx,n)
            
            return cnt>=k
        
        low=1
        high=m*n
        ans=low
        while low<=high:
            mid=low+(high-low)//2
            if count(mid)==True:
                ans=mid
                high=mid-1
            else:
                low=mid+1
        
        return ans



        